drivers/fuel_gauge/bq27z746/bq27z746.c

```diff
@@ -0,0 +1,324 @@
+
+static int bq27z746_read16(const struct device *dev, uint8_t reg, uint16_t *value)
+{
+	uint8_t i2c_data[2];
+	const struct bq27z746_config *cfg = dev->config;
+	const int status = i2c_burst_read_dt(&cfg->i2c, reg, i2c_data, sizeof(i2c_data));
+
+	if (status < 0) {
+		LOG_ERR("Unable to read register");
+		return status;
+	}
+	*value = sys_get_le16(i2c_data);
+
+	return 0;
+}
+
+static int bq27z746_write16(const struct device *dev, uint8_t reg, uint16_t value)
+{
+	uint8_t buf[3];
+	const struct bq27z746_config *cfg = dev->config;
+
+	buf[0] = reg;
+	sys_put_le16(value, &buf[1]);
+
+	return i2c_write_dt(&cfg->i2c, buf, sizeof(buf));
+}
+
+static int bq27z746_read_mac(const struct device *dev, uint16_t cmd, uint8_t *data, int len)
+{
+	if (len > BQ27Z746_MAC_DATA_LEN) {
+		return -EINVAL;
+	}
+
+	uint8_t buf[BQ27Z746_MAC_COMPLETE_LEN];
+	const struct bq27z746_config *cfg = dev->config;
+
+	/* Instead of MAC, ALTMAC is used as reccommended in the datasheet */
+	int ret = bq27z746_write16(dev, BQ27Z746_ALTMANUFACTURERACCESS, cmd);
+
+	if (ret != 0) {
+		return ret;
+	}
+
+	/*
+	 * The data read from BQ27Z746_ALTMANUFACTURERACCESS is:
+	 * 0..1: The command (for verification)
+	 * 2..33: The data
+	 * 34: Checksum calculated as (uint8_t)(0xFF - (sum of all command and data bytes))
+	 * 35: Length including command, checksum and length (e.g. data length + 4)
+	 */
+
+	ret = i2c_burst_read_dt(&cfg->i2c, BQ27Z746_ALTMANUFACTURERACCESS, buf,
+				BQ27Z746_MAC_COMPLETE_LEN);
+	if (ret != 0) {
+		return ret;
+	}
+
+	/* The first two bytes read is the command and is used for verification */
+	const uint16_t read_cmd = sys_get_le16(buf);
+
+	if (read_cmd != cmd) {
+		LOG_ERR("Read command 0x%x != written command 0x%x", read_cmd, cmd);
+		return -EIO;
+	}
+
+	const uint8_t checksum_actual = buf[34];
+	uint8_t sum = 0; /* Intentionally 8 bit wide and overflowing */
+
+	for (int i = 0; i < BQ27Z746_MAC_COMPLETE_LEN - 2; i++) {
+		sum += buf[i];
+	}
+
+	const uint8_t checksum_expected = 0xFF - sum;
+
+	if (checksum_expected != checksum_actual) {
+		LOG_ERR("Checksum mismatch");
+		return -EIO;
+	}
+
+	/* First byte of the user buffer is the length */
+	data[0] = buf[35] - BQ27Z746_MAC_OVERHEAD_LEN;
+
+	/* Copy only the data to the user buffer (= skipping the first two command bytes) */
+	memcpy(&data[1], &buf[2], len);
+
+	return ret;
+}
+
+static int bq27z746_get_prop(const struct device *dev, struct fuel_gauge_get_property *prop)
+{
+	int rc = 0;
+	uint16_t val = 0;
+
+	/*
+	 * Possibly negative values must be cast from uint16 to int16 first to
+	 * then correctly end up in the wider datatypes of `prop`.
+	 */
+
+	switch (prop->property_type) {
+	case FUEL_GAUGE_AVG_CURRENT:
+		rc = bq27z746_read16(dev, BQ27Z746_AVERAGECURRENT, &val);
+		prop->value.avg_current = (int16_t)val * 1000;
+		break;
+	case FUEL_GAUGE_CYCLE_COUNT:
+		rc = bq27z746_read16(dev, BQ27Z746_CYCLECOUNT, &val);
+		prop->value.cycle_count = val / 100;
```

Comment:
This register as I read the manual contains the whole number of cycle counts, meaning val * 100 is the correct output here.

A 1 would mean there has been 1/100th of a cycle (a partial charge cycle perhaps).

A more accurate output here might be cycle count percentage + (cycle count * 100) which bq27 has available.

Reply:
That is correct, thank you! Good catch, Tom!

However, looking back at it, I wonder about the actual utility of a tracking `1/100` of a cycle. I believe we made it `1/100` due as a request - I'll go look back at that. But that's a different discussion.

Reply:
@aaronemassey done

Reply:
@teburd you happy with the change?

Reply:
ping @teburd